### PR TITLE
build secp256 on android

### PR DESCRIPTION
**Pull request by debris**

### CODE REVIEW

File: build.rs [hunk 1]
```diff
@@ -23,28 +23,62 @@
...
+const ANDROID_INCLUDE: &'static str = "platforms/android-21/arch-arm64/usr/include";
+
+fn concat_paths(first: &str, second: &str) -> PathBuf {
+	let mut path = PathBuf::from(first);
+	path.push(second);
+	path
+}
+
+fn setup_android(config: &mut gcc::Config) {
+	let path = env::var_os("PATH").unwrap();
```
Comment (tomusdrw) [inline build.rs]: expect?

File: build.rs [hunk 2]
```diff
@@ -23,28 +23,62 @@
 
 extern crate gcc;
 
+use std::env;
+use std::path::PathBuf;
+
+const ANDROID_AARCH64_COMPILER: &'static str = "toolchains/aarch64-linux-android-4.9/prebuilt/darwin-x86_64/bin";
```
Comment (tomusdrw) [inline build.rs]: where do these paths come from?

File: build.rs [hunk 3]
```diff
@@ -23,28 +23,62 @@
...
+const ANDROID_ARM_COMPILER: &'static str = "toolchains/arm-linux-androideabi-4.9/prebuilt/darwin-x86_64/bin";
+const ANDROID_I686_COMPILER: &'static str = "toolchains/x86-4.9/prebuilt/darwin-x86_64/bin";
```
Comment (tomusdrw) [inline build.rs]: Might be good to extract `4.9` as a version and use `concat!`?

File: build.rs [hunk 4]
```diff
@@ -23,28 +23,62 @@
 
 extern crate gcc;
 
+use std::env;
+use std::path::PathBuf;
+
+const ANDROID_AARCH64_COMPILER: &'static str = "toolchains/aarch64-linux-android-4.9/prebuilt/darwin-x86_64/bin";
+const ANDROID_ARM_COMPILER: &'static str = "toolchains/arm-linux-androideabi-4.9/prebuilt/darwin-x86_64/bin";
```
Comment (tomusdrw) [inline build.rs]: What about other platforms? It will not really work on anything else the `darwin-x86_64` right?

File: build.rs [hunk 5]
```diff
@@ -23,28 +23,62 @@
...
+
+fn setup_android(config: &mut gcc::Config) {
+	let path = env::var_os("PATH").unwrap();
+	let ndk_home = env::var("NDK_HOME").expect("NDK_HOME is not set");
+	let mut paths = env::split_paths(&path).collect::<Vec<_>>();
+	paths.push(concat_paths(&ndk_home, ANDROID_AARCH64_COMPILER));
+	paths.push(concat_paths(&ndk_home, ANDROID_ARM_COMPILER));
+	paths.push(concat_paths(&ndk_home, ANDROID_I686_COMPILER));
+
+	let new_path = env::join_paths(paths).unwrap();
```
Comment (tomusdrw) [inline build.rs]: expect?

File: build.rs [hunk 6]
```diff
@@ -23,28 +23,62 @@
...
-    // secp256k1
-    base_config.file("depend/secp256k1/contrib/lax_der_parsing.c")
-	       .file("depend/secp256k1/src/ext.c")
-               .compile("libsecp256k1.a");
+	let mut base_config = gcc::Config::new();
+	base_config.include("depend/secp256k1/")
+		.include("depend/secp256k1/include")
+		.include("depend/secp256k1/src");
+
+	let target = env::var("TARGET").unwrap();
```
Comment (tomusdrw) [inline build.rs]: What if it's not set?

**Review by tomusdrw — COMMENTED** [review]
Grumbles

File: build.rs (hunk 6 above)
Comment (debris) [reply build.rs]: it's always set by `cargo` [docs](http://doc.crates.io/environment-variables.html)

File: build.rs (hunk 4 above)
Comment (debris) [reply build.rs]: `rustc` does not support any other platforms anyway

File: build.rs (hunk 2 above)
Comment (debris) [reply build.rs]: these are location of cross compilers installed by latest android-ndk for every target in

`multirust list-available-targets stable | grep android`

File: build.rs (hunk 3 above)
Comment (debris) [reply build.rs]: I cannot guarantee that other ndk versions will compile the code. New ndk versions are released 2-3 times a year. We can update the script then

File: build.rs (hunk 6 above)
Comment (debris) [reply build.rs]: I will add `expect`

File: build.rs [hunk 7]
```diff
@@ -23,28 +23,79 @@
...
+const OS: &'static str = "windows";
+
+const ANDROID_INCLUDE: &'static str = "platforms/android-21/arch-arm64/usr/include";
+
+fn android_aarch_compiler() -> String {
+	"toolchains/aarch64-linux-android-4.9/prebuilt/".to_owned() + OS + "-x86_64/bin"
+}
+
+fn android_arm_compiler() -> String {
+	"toolchains/arm-linux-androideabi-4.9/prebuilt/".to_owned() + OS + "-x86_64/bin"
```
Comment (tomusdrw) [inline build.rs:44]: `format!(` would proably look nicer

**Review by tomusdrw — APPROVED** [review]

File: build.rs (hunk 7 above)
Comment (debris) [reply build.rs:44]: I disagree. Also, it's less efficient, cause it needs to evaluate formatted expression

**Pull request merged**